src/nvim/tui/term_input.inl

```diff
@@ -0,0 +1,209 @@
+
+  while ((result = tk_getkey(input->tk, &key, force)) == TERMKEY_RES_KEY) {
+    if (key.type == TERMKEY_TYPE_UNICODE && !key.modifiers) {
+      forward_simple_utf8(&key);
+    } else if (key.type == TERMKEY_TYPE_UNICODE ||
+               key.type == TERMKEY_TYPE_FUNCTION ||
+               key.type == TERMKEY_TYPE_KEYSYM) {
+      forward_modified_utf8(input->tk, &key);
+    } else if (key.type == TERMKEY_TYPE_MOUSE) {
+      forward_mouse_event(input->tk, &key);
+    }
+  }
+
+  if (result != TERMKEY_RES_AGAIN) {
+    return;
+  }
+
+  Integer ms = 0;
+  bool timeout = false;
+  // Check 'timeout' and 'ttimeout' to determine if we should send ESC
+  // after 'ttimeoutlen'. See :help 'ttimeout' for more information
+  Error err;
+  timeout = vim_get_option(cstr_as_string("timeout"), &err).data.boolean;
+  if (!timeout) {
+    timeout = vim_get_option(cstr_as_string("ttimeout"), &err).data.boolean;
+  }
+
+  if (timeout) {
+    ms = vim_get_option(cstr_as_string("ttimeoutlen"), &err).data.integer;
+  }
+
+  if (ms > 0) {
+    // Stop the current timer if already running
+    uv_timer_stop(&input->timer_handle);
+    uv_timer_start(&input->timer_handle, timer_cb, (uint32_t)ms, 0);
+  } else {
+    tk_getkeys(input, true);
+  }
+}
+
+static void timer_cb(uv_timer_t *handle)
+{
+  tk_getkeys(handle->data, true);
+}
+
+static void read_cb(RStream *rstream, void *rstream_data, bool eof)
+{
+  if (eof) {
+    input_done();
+    return;
+  }
+
+  TermInput *input = rstream_data;
+
+  do {
+    size_t consumed = termkey_push_bytes(input->tk,
+        rbuffer_read_ptr(input->read_buffer),
+        rbuffer_pending(input->read_buffer));
+    rbuffer_consumed(input->read_buffer, consumed);
+    tk_getkeys(input, false);
+  } while (rbuffer_pending(input->read_buffer));
+}
+
+static TermInput *term_input_new(void)
+{
+  TermInput *rv = xmalloc(sizeof(TermInput));
+  // read input from stderr if stdin is not a tty
+  rv->in_fd = isatty(0) ? 0 : (isatty(2) ? 2 : 0);
+  rv->tk = termkey_new_abstract(os_getenv("TERM"), 0);
+  int curflags = termkey_get_canonflags(rv->tk);
+  termkey_set_canonflags(rv->tk, curflags | TERMKEY_CANON_DELBS);
+  // setup input handle
+  uv_tty_init(uv_default_loop(), &rv->input_handle, rv->in_fd, 1);
+  uv_tty_set_mode(&rv->input_handle, UV_TTY_MODE_RAW);
+  rv->input_handle.data = NULL;
+  rv->read_buffer = rbuffer_new(0xfff);
+  rv->read_stream = rstream_new(read_cb, rv->read_buffer, rv);
+  rstream_set_stream(rv->read_stream, (uv_stream_t *)&rv->input_handle);
+  rstream_start(rv->read_stream);
+  // initialize a timer handle for handling ESC with libtermkey
+  uv_timer_init(uv_default_loop(), &rv->timer_handle);
+  rv->timer_handle.data = rv;
+  return rv;
+}
+
+static void term_input_destroy(TermInput *input)
+{
+  uv_tty_reset_mode();
+  uv_timer_stop(&input->timer_handle);
+  rstream_stop(input->read_stream);
+  rstream_free(input->read_stream);
+  uv_close((uv_handle_t *)&input->input_handle, NULL);
```

Comment:
Member `free_handle` is initialized with `false` for an `RStream`. If it is not changed, `rstream_free()` would not call `close_cb()`.